### PR TITLE
[Merge-Queue] Add queue triggered by label addition

### PR DESCRIPTION
**Pull request by JonWBedard**
#### 7f4060ec700e3303a4559f3362e7adacde48b1eb
<pre>
[Merge-Queue] Add queue triggered by label addition
<a href="https://bugs.webkit.org/show_bug.cgi?id=237615">https://bugs.webkit.org/show_bug.cgi?id=237615</a>
&lt;rdar://problem/89983452 &gt;

Reviewed by Aakash Jain.

* Tools/CISupport/ews-build/config.json: Add Merge-Queue.
* Tools/CISupport/ews-build/events.py:
(GitHubEventHandlerNoEdits): Remove action list, add merge-queue labels.
(GitHubEventHandlerNoEdits.handle_pull_request): Let parent class filter actions, treat
merge-queue label addition as a &apos;synchronize&apos;, but change event name to trigger different
scheduler family.
* Tools/CISupport/ews-build/factories.py:
(MergeQueueFactory): Added.
* Tools/CISupport/ews-build/factories_unittest.py:
(TestExpectedBuildSteps):
* Tools/CISupport/ews-build/loadConfig.py:
(loadBuilderConfig): AnyBranchSchedulers should only accept events which match their name.
* Tools/CISupport/ews-build/loadConfig_unittest.py:
(ConfigDotJSONTest): Match pull request scheduler name to pull request event.

Canonical link: <a href="https://commits.webkit.org/248349@main">https://commits.webkit.org/248349@main</a>
git-svn-id: <a href="https://svn.webkit.org/repository/webkit/trunk@291193">https://svn.webkit.org/repository/webkit/trunk@291193</a> 268f45cc-cd09-0410-ab3c-d52691b4dbfc
</pre>

### CODE REVIEW

**Comment by JonWBedard** [comment]
https://ews-build.webkit-uat.org/#/builders/85/builds/7 is a good example of Merge-Queue being triggered, while the checks on this PR are a good example of Merge-Queue not being triggered by normal PR interactions

File: Tools/CISupport/ews-build/config.json [hunk 1]
```diff
@@ -347,7 +347,7 @@
       "builderNames": ["Commit-Queue"]
     },
     {
-      "type": "AnyBranchScheduler", "name": "pullrequest",
+      "type": "AnyBranchScheduler", "name": "pull_request",
```
Comment (aj062) [inline Tools/CISupport/ews-build/config.json]: why this change? Is this name special?

File: Tools/CISupport/ews-build/events.py [hunk 2]
```diff
@@ -311,8 +311,9 @@ def stopService(self):
 
 
 class GitHubEventHandlerNoEdits(GitHubEventHandler):
-    ACTIONS_TO_TRIGGER_EWS = ('opened', 'synchronize')
```
Comment (aj062) [inline Tools/CISupport/ews-build/events.py]: Why is this being removed?

Comment (jain-aakash) [inline Tools/CISupport/ews-build/events.py:314]: Why is this being removed?

File: Tools/CISupport/ews-build/loadConfig.py [hunk 3]
```diff
@@ -80,10 +81,13 @@ def loadBuilderConfig(c, is_test_mode_enabled=False, master_prefix_path='./'):
     c['schedulers'] = []
     for scheduler in config['schedulers']:
         schedulerClassName = scheduler.pop('type')
+        schedulerName = scheduler.get('name')
         schedulerClass = globals()[schedulerClassName]
         if (schedulerClassName == 'Try_Userpass'):
             # FIXME: Read the credentials from local file on disk.
             scheduler['userpass'] = [(passwords.get('BUILDBOT_TRY_USERNAME', 'sampleuser'), passwords.get('BUILDBOT_TRY_PASSWORD', 'samplepass'))]
+        if schedulerClassName == 'AnyBranchScheduler' and schedulerName:
+            scheduler['change_filter'] = ChangeFilter(filter_fn=lambda c, schedulerName=schedulerName: c.properties.getProperty('event') == schedulerName)
```
Comment (aj062) [inline Tools/CISupport/ews-build/loadConfig.py]: Can we please have a non-lambda version for better readability?

Comment (jain-aakash) [inline Tools/CISupport/ews-build/loadConfig.py:90]: Can we please have a non-lambda version for better readability?

File: Tools/CISupport/ews-build/events.py [hunk 4]
```diff
@@ -311,8 +311,9 @@ def stopService(self):
...
+    LABELED = 'labeled'
```
Comment (aj062) [inline Tools/CISupport/ews-build/events.py]: this variable name seems weird

Comment (jain-aakash) [inline Tools/CISupport/ews-build/events.py:314]: this variable name seems weird

File: Tools/CISupport/ews-build/config.json (hunk 1 above)
Comment (JonWBedard) [reply Tools/CISupport/ews-build/config.json]: When I first added this scheduler, I thought the name was special. It was not. But with this change, the name of a scheduler needs to match the even which is triggering it. For `merge_queue`, it would look like this:
```
{
    "type": "AnyBranchScheduler", "name": "merge_queue",
    "builderNames": ["Merge-Queue"]
}
```

File: Tools/CISupport/ews-build/events.py (hunk 2 above)
Comment (JonWBedard) [reply Tools/CISupport/ews-build/events.py]: Turns out `GitHubEventHandler` already handles this for us, and is the reason we need to lie about the action for labeling on line 362

File: Tools/CISupport/ews-build/events.py (hunk 4 above)
Comment (JonWBedard) [reply Tools/CISupport/ews-build/events.py]: Would you prefer we reference it as a string literal? Change the name to `LABELED_ACTION`?

Comment (JonWBedard) [reply Tools/CISupport/ews-build/events.py]: Just removed the variable entirely.

File: Tools/CISupport/ews-build/loadConfig.py [hunk 5]
```diff
@@ -80,10 +81,17 @@ def loadBuilderConfig(c, is_test_mode_enabled=False, master_prefix_path='./'):
     c['schedulers'] = []
     for scheduler in config['schedulers']:
         schedulerClassName = scheduler.pop('type')
+        schedulerName = scheduler.get('name')
         schedulerClass = globals()[schedulerClassName]
+
+        def filter_fn(c, schedulerName=schedulerName):
```
Comment (aj062) [inline Tools/CISupport/ews-build/loadConfig.py]: Nit: c=>change

File: Tools/CISupport/ews-build/events.py [hunk 6]
```diff
@@ -350,10 +350,14 @@ def handle_pull_request(self, payload, event):
...
         state = payload.get('pull_request', {}).get('state')
-        if action not in self.ACTIONS_TO_TRIGGER_EWS:
-            log.msg('Action {} on PR #{} does not indicate code has been changed'.format(action, pr_number))
-            return ([], 'git')
+        labels = [label.get('name') for label in payload.get('pull_request', {}).get('labels', [])]
+
         if state not in self.OPEN_STATES:
             log.msg("PR #{} is '{}', which triggers nothing".format(pr_number, state))
             return ([], 'git')
+        if action == 'labeled' and any(label in self.MERGE_QUEUE_LABELS for label in labels):
```
Comment (aj062) [inline Tools/CISupport/ews-build/events.py]: what happens if the label was removed?

File: Tools/CISupport/ews-build/events.py (hunk 2 above)
Comment (aj062) [reply Tools/CISupport/ews-build/events.py]: Seems like buildbot is also checking for reopened:
reference:https://github.com/buildbot/buildbot/blob/master/master/buildbot/www/hooks/github.py#L190

File: Tools/CISupport/ews-build/events.py (hunk 6 above)
Comment (JonWBedard) [reply Tools/CISupport/ews-build/events.py]: Action is `unlabeled` in that case

File: Tools/CISupport/ews-build/events.py (hunk 2 above)
Comment (JonWBedard) [reply Tools/CISupport/ews-build/events.py]: Yes, which is seems desirable to me, since closing a PR will stop further processing

**Review by aj062 — APPROVED** [review]

**Comment by JonWBedard** [comment]
Landed https://commits.webkit.org/248349@main (7f4060ec700e)!
